sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,422 @@
+	return r.class
+}
+
+// Permission returns the opaque permission that is being requested.
+func (r *Request) Permission() interface{} {
+	return r.permission
+}
+
+// Reply sends the given response back to the kernel.
+func (r *Request) Reply(response interface{}) error {
+	if !atomic.CompareAndSwapUint32(&r.replied, 0, 1) {
+		return ErrAlreadyReplied
+	}
+	r.replyChan <- response
+	return nil
+}
+
+// Listener encapsulates a loop for receiving apparmor notification requests
+// and responding with notification responses, hiding the low-level details.
+type Listener struct {
+	// reqs is a channel with incoming requests. Each request is asynchronous
+	// and needs to be replied to.
+	reqs chan *Request
+
+	notifyFile *os.File
+	poll       *epoll.Epoll
+
+	tomb tomb.Tomb
+
+	// status keeps track of whether the listener has been run and/or closed,
+	// both to ensure that Run() and Close() are executed at most once each,
+	// and to ensure that the listener cannot be run after it has been closed.
+	// Must be read/modified atomically, and can only be changed in one of the
+	// following ways:
+	// - statusReady -> statusRunning
+	// - statusReady -> statusClosed
+	// - statusRunning -> statusClosed
+	status uint32
+}
+
+const (
+	statusReady uint32 = iota
+	statusRunning
+	statusClosed
+)
+
+// Register opens and configures the apparmor notification interface.
+//
+// If the kernel does not support the notification mechanism the error is ErrNotSupported.
+func Register() (listener *Listener, err error) {
+	path := notify.SysPath
+	if override := os.Getenv("PROMPT_NOTIFY_PATH"); override != "" {
+		path = override
+	}
+
+	notifyFile, err := osOpen(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+	defer func() {
+		if err != nil {
+			notifyFile.Close()
+		}
+	}()
+
+	msg := notify.MsgNotificationFilter{ModeSet: notify.APPARMOR_MODESET_USER}
+	data, err := msg.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	ioctlBuf := notify.IoctlRequestBuffer(data)
+	_, err = notifyIoctl(notifyFile.Fd(), notify.APPARMOR_NOTIF_SET_FILTER, ioctlBuf)
+	if err != nil {
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			poll.Close()
+		}
+	}()
+	if err = poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener = &Listener{
+		reqs: make(chan *Request, 1),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+// Close stops the listener and closes the kernel communication file.
+// Returns once all waiting goroutines terminate and the communication file
+// is closed.
+func (l *Listener) Close() error {
+	l.tomb.Kill(nil)
```

Comment:
Good catch! Originally, `l.tomb.Wait()` was called in one of the cases, but that has since been moved to after the switch. I think it makes more sense to move it down to after the switch, though before the `l.poll.Close()`, so we explicitly make sure the tomb dies with the error we want, which should be `ErrClosed` --- `nil` is incorrect here. `l.poll.Close()` currently causes `ErrClosed` to be sent elsewhere, but may as well do so explicitly here.